pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -0,0 +1,279 @@
+		#[pallet::constant]
+		type MaxCurrencies: Get<u32> + TypeInfo;
+		/// The deposit required for each bonded currency.
+		#[pallet::constant]
+		type DepositPerCurrency: Get<DepositCurrencyBalanceOf<Self>>;
+		/// Who can create new bonded currency pools.
+		type PoolCreateOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Self::AccountId>;
+		/// The type used for pool ids
+		type PoolId: Parameter
+			+ MaxEncodedLen
+			+ From<[u8; 32]>
+			+ Into<Self::AccountId>
+			+ Into<DepositCurrencyHoldReasonOf<Self>>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	/// Bonded Currency Swapping Pools
+	#[pallet::storage]
+	#[pallet::getter(fn pools)]
+	pub(crate) type Pools<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		T::PoolId,
+		PoolDetails<T::AccountId, FungiblesAssetIdOf<T>, Curve, T::MaxCurrencies>,
+		OptionQuery,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A new bonded token pool has been initiated. [pool_id]
+		PoolCreated(T::AccountId),
+		/// Trading locks on a pool have been removed. [pool_id]
+		Unlocked(T::AccountId),
+		/// Trading locks on a pool have been set or changed. [pool_id]
+		LockSet(T::AccountId),
+		/// A bonded token pool has been moved to destroying state. [pool_id]
+		DestructionStarted(T::AccountId),
+		/// A bonded token pool has been fully destroyed and all collateral and deposits have been refunded. [pool_id]
+		Destroyed(T::AccountId),
+	}
+
+	// Errors inform users that something went wrong.
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The number of bonded currencies on a new pool is either lower than 1 or greater than MaxCurrencies.
+		CurrenciesOutOfBounds,
+		/// A token swap cannot be executed due to a lock placed on this operation.
+		Locked,
+		/// The pool id is not currently registered.
+		PoolUnknown,
+		/// The pool has no associated bonded currency with the given index.
+		IndexOutOfBounds,
+		/// The cost or returns for a mint, burn, or swap operation is outside the user-defined slippage tolerance.
+		Slippage,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		FungiblesBalanceOf<T>: TryInto<CollateralCurrencyBalanceOf<T>>,
+	{
+		#[pallet::call_index(0)]
+		// #[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))] TODO: properly configure weights
+		pub fn create_pool(
+			origin: OriginFor<T>,
+			curve: Curve,
+			currencies: BoundedVec<TokenMeta<FungiblesBalanceOf<T>, FungiblesAssetIdOf<T>>, T::MaxCurrencies>,
+			frozen: bool,
+			// currency_admin: Option<AccountIdLookupOf<T>> TODO: use this to set currency admin
+		) -> DispatchResultWithPostInfo {
+			// ensure origin is PoolCreateOrigin
+			let who = T::PoolCreateOrigin::ensure_origin(origin)?;
+
+			ensure!(
+				(2..=T::MaxCurrencies::get().saturated_into()).contains(&currencies.len()),
```

Comment:
Oh this is a solution that still stuck around from the previous designs. Should be 
```suggestion
				(1..=T::MaxCurrencies::get().saturated_into()).contains(&currencies.len()),
```
as well. It's a suggestion by Antonio, bc we also need to check that the length is not 0 (larger than 1 previously) - but we can probably also check that differently now?

Reply:
Is it an issue, if no currencies are set? The user can directly delete it. It does not really make sense but who cares :grimacing:

Reply:
it's super easy to prevent, it's an edge case with no utility (which may bite us in the ass later if we don't check it), so I don't see why we would choose to allow creating pools with no tokens attached?

Reply:
range fixed in 771bb33ede3c47260cf81f4b0512b051aba9931e